feat(zone.js): upgrade zone.js to angular package format(APF)

JiaLiPassion opened the pull request:
Close #35157

In the current version of zone.js, zone.js uses it's own package format, and it is not following the rule
of Angualr package format(APF), so it is not easily to be consumed by Angular CLI or other bundle tools.
For example, zone.js npm package has two bundles,

1. zone.js/dist/zone.js, this is a `es5` bundle.
2. zone.js/dist/zone-evergreen.js, this is a `es2015` bundle.

And Angular CLI has to add some hard-coding code to handle this case, ohttps://github.com/angular/angular-cli/blob/5376a8b1392ac7bd252782d8474161ce03a4d1cb/packages/schematics/angular/application/files/src/polyfills.ts.template#L55-L58

This PR upgrade zone.js npm package format to follow APF rule, https://docs.google.com/document/d/1CZC2rcpxffTDfRDs6p1cfbmKNLA6x5O-NtkJglDaBVs/edit#heading=h.k0mh3o8u5hx

The updated points are:

1. in package.json, update all bundle related properties

```
  "main": "./bundles/zone.umd.js",
  "module": "./fesm2015/zone.js",
  "es2015": "./fesm2015/zone.js",
  "fesm2015": "./fesm2015/zone.js",
```

2. re-organize dist folder, for example for `zone.js` bundle, now we have

```
  dist/
      bundles/
             zone.js            // this is the es5 bundle
      fesm2015/
             zone.js            // this is the es2015 bundle (in the old version is `zone-evergreen.js`)
```

3. have several sub-packages.

1. `zone-testing`, provide zone-testing bundles include zone.js and testing libraries
2. `zone-node`, provide zone.js implemention for NodeJS
3. `zone-mix`, provide zone.js patches for both Browser and NodeJS

All those sub-packages will have their own `package.json` and the bundle will reference `bundles(es5)` and `fesm2015(es2015)`.

4. `typings` property are removed from `package.json`, instead using `files` property, because now we have several `d.ts` files as typings.

```
"files": [
    "./zone.js.d.ts",
    "./zone.api.extensions.ts",
    "./zone.configurations.api.ts"
  ],
```

## Does this PR introduce a breaking change?

- [x] Yes
- [ ] No

All `zone.js` npm package folder and file name are changed, so we need to 

1.  create PR for angular/cli to migrate the changes.
2. make announcement for the users not using angular/cli. 

I created a sample repo to show the contents of the new npm package contents.
https://github.com/JiaLiPassion/zone-new-package-format/tree/master/npm_package

1. `package.json`, https://github.com/JiaLiPassion/zone-new-package-format/blob/master/npm_package/package.json
2. `bundles folder`, https://github.com/JiaLiPassion/zone-new-package-format/tree/master/npm_package/bundles
3. `fesm2015 folder`, https://github.com/JiaLiPassion/zone-new-package-format/tree/master/npm_package/bundles
4. `node sub package`, https://github.com/JiaLiPassion/zone-new-package-format/tree/master/npm_package/node
5. `mix sub package`, https://github.com/JiaLiPassion/zone-new-package-format/tree/master/npm_package/mix
6. `testing`, https://github.com/JiaLiPassion/zone-new-package-format/tree/master/npm_package/testing

Please check whether the output satisfy the requirements.

And the following up tasks will be.

1. make `angular repo` work with the new format.
2. add PR for `angular cli` for migration.
3. update documentation involving the `import 'zone.js/dist/zone'` of `polyfills.ts`.

IgorMinar commented on packages/zone.js/lib/zone-spec/fake-async-test.ts:
```diff
@@ -30,29 +30,28 @@ interface MacroTaskOptions {
-      const args = Array.prototype.slice.call(arguments);
-      return new OriginalDate(...args);
-    }
+function FakeDate() {
```
if this is intentionally manually downleveled to es5 we should leave a comment here explaining why and when it's safe to switch back to `class` syntax

JiaLiPassion replied:
Got it, I will add the comment. Thanks!

JiaLiPassion replied:
Added comments to explain why change `class` to `function` since we now compile this file also to `es2015` (which we didn't do before this PR), and in `es2015`, the `prototype` of a class is `readonly`, and will cause compile error since we need to assign `FakeDate.prototype = Date.prototype`.

I will come back for another PR to find another solution instead of modifying `FakeDate.prototype` directly.

IgorMinar reviewed (changes requested):
this looks quite good! thanks @JiaLiPassion!

Can you please do me a favor a fix one big issue: in order not to make this a breaking change, can you please ensure that the built npm package contains `dist/package.json` which redirects the tools trying to resolve `zone.js/dist/zone` to the right place? essentially you need to resolver redirect from`zone.js/dist/zone` to `zone.js/`.

This will ensure that all Angular CLI apps remain operational, in spite of them having the following in `polyfills.ts`

```
/***************************************************************************************************
 * Zone JS is required by default for Angular itself.
 */
import 'zone.js/dist/zone';  // Included with Angular CLI.
```

JiaLiPassion commented:
@IgorMinar , thank you for the review.

> Can you please do me a favor a fix one big issue: in order not to make this a breaking change, can you please ensure that the built npm package contains dist/package.json which redirects the tools trying to resolve `zone.js/dist/zone` to the right place? essentially you need to resolver redirect from `zone.js/dist/zone` to `zone.js/.`

To do that, in my understanding I need to output additional bundles to `dist/` folder, since current in the `polyfill.ts`, the code is 

```
import 'zone.js/dist/zone';  // Included with Angular CLI.
```

So it expect that there is a `zone.js` file under `node_modules/zone.js/dist/` folder, if the import is `import 'zone.js'`, I think I can change the `main/browser` field in `package.json` to do the redirect, but since now the import is the specified file, I still need to put all bundles under the `zone.js/dist` folder, is my understanding correct?

IgorMinar commented:
Nope. I know that this is tricky because it's not common to have multiple package.jsons in a single package. But that is really what I'm asking for in this case. Ommitting all the other files the package should have these two package.jsons:


/package.json
/dist/zone/package.json

The first one will be used by yarn/npm and should contain all the normal fields. the second one will be used just by webpack (and other build tools) and will redirect the resolution of imports to "zone.js/dist/zone" to the new file location.

Does that make sense?

JiaLiPassion commented:
@IgorMinar , thank you for the detailed instruction, I now know how to do it, I will continue to work on this.

IgorMinar commented:
Sounds great. Thanks!

JiaLiPassion commented:
@IgorMinar, I updated the PR and add all `secondary entry point` for backward compatibility.
There are still breaking changes, such as 
https://github.com/angular/angular/pull/36540/files#diff-180e981fb73a71e2b70ae7be7f067931R15
or 
https://github.com/angular/angular/pull/36540/files#diff-706fde4cbcd621568eef9bd11996ff60R9

In these cases, the application will reference or require directly the `js` file from `.node_modules`. So after this version of zone.js is release, those cases need to update their applications.

Please review, thank you.

IgorMinar commented on packages/zone.js/dist/async-test/BUILD.bazel:
```diff
@@ -0,0 +1,3 @@
+exports_files([
```
I *think* that you don't actually need the BUILD.bazel file here at all because as long as you are trying to get hold of a file from within the same bazel package, you should be able to just use relative labels `:async-test.min/package.json` from within packages/zone.js/dist/BUILD.bazel

JiaLiPassion replied:
Got it, thanks for the tips, I have updated the bazel files.

IgorMinar reviewed (changes requested):
this seems to be on the right track.. a few more changes requested. Thanks @JiaLiPassion !

JiaLiPassion committed:
feat(zone.js): upgrade zone.js to angular package format(APF)

Close #35157

In the current version of zone.js, zone.js uses it's own package format, and it is not following the rule
of Angualr package format(APF), so it is not easily to be consumed by Angular CLI or other bundle tools.
For example, zone.js npm package has two bundles,

1. zone.js/dist/zone.js, this is a `es5` bundle.
2. zone.js/dist/zone-evergreen.js, this is a `es2015` bundle.

And Angular CLI has to add some hard-coding code to handle this case, ohttps://github.com/angular/angular-cli/blob/5376a8b1392ac7bd252782d8474161ce03a4d1cb/packages/schematics/angular/application/files/src/polyfills.ts.template#L55-L58

This PR upgrade zone.js npm package format to follow APF rule, https://docs.google.com/document/d/1CZC2rcpxffTDfRDs6p1cfbmKNLA6x5O-NtkJglDaBVs/edit#heading=h.k0mh3o8u5hx

The updated points are:

1. in package.json, update all bundle related properties

```
  "main": "./bundles/zone.umd.js",
  "module": "./fesm2015/zone.js",
  "es2015": "./fesm2015/zone.js",
  "fesm2015": "./fesm2015/zone.js",
```

2. re-organize dist folder, for example for `zone.js` bundle, now we have

```
  dist/
      bundles/
             zone.js            // this is the es5 bundle
      fesm2015/
             zone.js            // this is the es2015 bundle (in the old version is `zone-evergreen.js`)
```

3. have several sub-packages.

1. `zone-testing`, provide zone-testing bundles include zone.js and testing libraries
2. `zone-node`, provide zone.js implemention for NodeJS
3. `zone-mix`, provide zone.js patches for both Browser and NodeJS

All those sub-packages will have their own `package.json` and the bundle will reference `bundles(es5)` and `fesm2015(es2015)`.

4. keep backward compatibility, still keep the `zone.js/dist` folder, and all bundles will be redirected to `zone.js/bundles` or `zone.js/fesm2015` folders.

JiaLiPassion commented on integration/bazel-schematics/package.json:
```diff
@@ -4,32 +4,33 @@
-    "@angular/platform-browser-dynamic": "file:../../dist/packages-dist/platform-browser-dynamic",
-    "@angular/router": "file:../../dist/packages-dist/router",
-    "@types/node": "file:../../node_modules/@types/node",
+    "@angular-devkit/build-angular": "file:/private/var/tmp/_bazel_lijia/5cf5316747fa3605a63b048653aecc47/execroot/angular/bazel-out/darwin-fastbuild/bin/integration/bazel-schematics_test.debug.sh.runfiles/npm/angular-devkit_build-angular_archive.tar.gz",
```
Got it, sorry not sure why this file is changed, I have reverted it.

JiaLiPassion commented on integration/bazel/package.json:
```diff
@@ -4,49 +4,49 @@
   "license": "MIT",
   "dependencies": {
-    "@angular/animations": "file:../../dist/packages-dist/animations",
+    "@angular/animations": "file:/private/var/tmp/_bazel_lijia/5cf5316747fa3605a63b048653aecc47/execroot/angular/bazel-out/darwin-fastbuild/bin/integration/bazel_test.debug.sh.runfiles/angular/packages/animations/npm_package_archive.tar.gz",
```
Got it, sorry not sure why this file is changed, I have reverted it.

JiaLiPassion commented on integration/platform-server/package.json:
```diff
@@ -8,35 +8,35 @@
-    "@angular/platform-browser": "file:../../dist/packages-dist/platform-browser",
-    "@angular/platform-browser-dynamic": "file:../../dist/packages-dist/platform-browser-dynamic",
-    "@angular/platform-server": "file:../../dist/packages-dist/platform-server",
+    "@angular/animations": "file:/private/var/tmp/_bazel_lijia/5cf5316747fa3605a63b048653aecc47/execroot/angular/bazel-out/darwin-fastbuild/bin/integration/platform-server_test.debug.sh.runfiles/angular/packages/animations/npm_package_archive.tar.gz",
```
Got it, sorry not sure why this file is changed, I have reverted it.

JiaLiPassion commented on integration/hello_world__closure/package.json:
```diff
@@ -3,30 +3,30 @@
-    "@angular/core": "file:../../dist/packages-dist/core",
-    "@angular/platform-browser": "file:../../dist/packages-dist/platform-browser",
-    "@angular/platform-server": "file:../../dist/packages-dist/platform-server",
+    "@angular/animations": "file:/private/var/tmp/_bazel_lijia/5cf5316747fa3605a63b048653aecc47/execroot/angular/bazel-out/darwin-fastbuild/bin/integration/hello_world__closure_test.debug.sh.runfiles/angular/packages/animations/npm_package_archive.tar.gz",
```
Got it, sorry not sure why this file is changed, I have reverted it.

JiaLiPassion commented on packages/zone.js/dist/async-test.min/package.json:
```diff
@@ -0,0 +1,7 @@
+{
+  "name": "async-test.min",
```
Got it, updated all names.

IgorMinar reviewed (commented):
this change looks great now! @JiaLiPassion can you please rebase it and resolve the conflicts and we can then get it in.

I'll kick of a presubmit on the current version just so that we know if there are any issues caused by this in google3. I don't expect that there will be any.

IgorMinar commented:
@JiaLiPassion please rerequest a review once the PR is rebased.

JiaLiPassion commented:
@IgorMinar , thank you for the review, I have rebased the PR.

JiaLiPassion committed:
feat(zone.js): move all zone optional bundles to plugins folders

Zone.js has a lot of optional bundles, such as `zone-patch-message-port`, those
bundles are monkey patch for specified APIs usually for soem experimental APIs or
some old APIs only available for specified platforms. Those bundles will not be
loaded by default.

In this commit, since we have several main `sub packages` such as `zone`, `zone-node`,
`zone-testing`, I put all the optional bundles under `plugins` folders for consistency.

IgorMinar commented:
I rebased this PR and force pushed because the PR was red due to being too far behind the master@HEAD. Hopefully the CI will now go green.

JiaLiPassion commented:
@IgorMinar, I just rebased again, now the CI is green, thanks.

IgorMinar reviewed (approved):
LGTM! Thank you!

Reviewed-for: global-approvers
